src/geometry.rs

```diff
@@ -1,9 +1,115 @@
 //! Extension's to Smithay's geometry module.
 
-use std::cmp;
+use std::error::Error;
+use std::str::FromStr;
+use std::{cmp, fmt, ops};
 
 use smithay::utils::{Point, Size};
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Matrix3x3<T: Copy> {
+    storage: Vec<T>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct FromVecError(usize, usize);
+
+impl fmt::Display for FromVecError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "Mismatched size when createing object from vector, expected {} got {}",
+            self.0, self.1
+        )
+    }
+}
+
+impl Error for FromVecError {}
+
+impl<T: Copy> TryFrom<Vec<T>> for Matrix3x3<T> {
+    type Error = Box<dyn Error>;
+
+    fn try_from(storage: Vec<T>) -> Result<Self, Self::Error> {
+        if storage.len() != 9 {
+            Err(Box::new(FromVecError(9, storage.len())))
+        } else {
+            Ok(Self { storage })
+        }
+    }
+}
+
+impl ops::Mul<Vector3D<f32>> for &Matrix3x3<f32> {
+    type Output = Vector3D<f32>;
+
+    fn mul(self, rhs: Vector3D<f32>) -> Self::Output {
+        let x = self.storage[0] * rhs.x + self.storage[1] * rhs.y + self.storage[2] * rhs.z;
+        let y = self.storage[3] * rhs.x + self.storage[4] * rhs.y + self.storage[5] * rhs.z;
+        let z = self.storage[6] * rhs.x + self.storage[7] * rhs.y + self.storage[8] * rhs.z;
+        Self::Output { x, y, z }
+    }
+}
+
+// The accepted formats is is 0, 0, 0; 0, 0, 0; 0, 0, 0.
+impl<T: FromStr + Copy> FromStr for Matrix3x3<T>
+where
+    <T as FromStr>::Err: Error + 'static,
+{
+    type Err = Box<dyn Error>;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let rows = s
+            .split(';')
+            // Ensure that we process only 3 rows.
+            .take(3)
+            // Take only 3 element from each row.
+            .flat_map(|row| row.split(',').take(3))
+            .map(|ch| T::from_str(ch.trim()))
+            .collect::<Result<Vec<T>, _>>()?;
+
+        Self::try_from(rows)
+    }
+}
+
+/// A point in the 3 dimensional space.
+#[derive(Debug, Clone, Copy, PartialEq, Default)]
+pub struct Vector3D<T: Copy> {
+    pub x: T,
+    pub y: T,
+    pub z: T,
+}
+
+impl<T: Copy> Vector3D<T> {
+    #[inline]
+    pub fn new(x: T, y: T, z: T) -> Self {
+        Self { x, y, z }
+    }
+}
```

Comment:
```suggestion
impl Vector3D {
    #[inline]
    pub fn new(x: T, y: T, z: T) -> Self {
        Self { x, y, z }
    }
}
```

I don't think any of these trait impl restrictions are actually necessary?